qutebrowser/completion/strategies/configmodel.py

```diff
@@ -0,0 +1,194 @@
+
+# Copyright 2014-2021 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Functions that return config-related completion models."""
+
+from typing import Optional, Callable, Any, Tuple
+
+from qutebrowser.completion.strategies.strategy import CompletionStrategy
+from qutebrowser.config import configdata, configexc
+from qutebrowser.completion import completionmodel, util
+from qutebrowser.completion.categories import listcategory
+from qutebrowser.commands import runners, cmdexc
+from qutebrowser.keyinput import keyutils
+from qutebrowser.completion.completer import CompletionInfo
+from qutebrowser.completion.completionmodel import CompletionModel
+
+
+class OptionMaker(CompletionStrategy):
+    """A CompletionModel that is generated for several option sets.
+
+    Args:
+        info: The config info that can be passed through.
+        title: The title of the options.
+        predicate: The function for filtering out the options. Takes a single
+                   argument.
+    """
+
+    COLUMN_WIDTHS = (20, 70, 10)
+
+    def __init__(self, title: str, predicate: Optional[Callable[[Optional[Any]], bool]] = None):
+        super().__init__()
+        self.title = title
+        self.predicate = predicate
+
+    def populate(self, *args: str, info: Optional[CompletionInfo]) -> CompletionModel:
+        super().populate(*args, info=info)
+        options = ((opt.name, opt.description, info.config.get_str(opt.name))
+                   for opt in configdata.DATA.values()
+                   if self.predicate(opt))
+        self.model.add_category(listcategory.ListCategory(self.title, options))
+        return self.model
+
+
+class Option(OptionMaker):
+    """A CompletionModel filled with settings and their descriptions."""
+
+    def __init__(self):
+        super().__init__("Options", lambda opt: not opt.no_autoconfig)
+
+
+class CustomizedOption(CompletionStrategy):
+    """A CompletionModel filled with set settings and their descriptions."""
+
+    COLUMN_WIDTHS = (20, 70, 10)
+
+    def populate(self, *args: str, info: Optional[CompletionInfo]) -> CompletionModel:
+        super().populate(*args, info=info)
+        options = ((values.opt.name, values.opt.description, info.config.get_str(values.opt.name))
+                   for values in info.config
```

Comment:
Yup. Makes sense. 
Have a question regarding your other comment about overriding title and predicate (I can't reply to that for some reason). I agree with the predicate part. But regarding the title part, should I add some sort of default in a similar way to column widths? And what would the default be? Asking this because, I don't think there's a way to force subclasses to override like the way the AbstractClass you described works with pylint.

Reply:
Actually regarding _option_source, the problem I have is by default it is returns an iterator of opt. But for CustomizedOption, info.config is an iterator of values containing an opt. This makes it quite difficult. And I don't know if doing any iteration in the _option_source would be a good idea. There's also the matter of predicate which is different for CustomizedOption because it doesn't take opt.